ethereum/slogging.py

```diff
@@ -138,11 +85,228 @@ def pop_records(self):
+
+# helpers
+known_loggers = set()  # know to structlog (i.e. maybe not yet initialized w/ logging)
+
+
+def get_logger_names():
+    # logging.Logger.manager.loggerDict.keys() # used ones
+    # initialized at module load get_logger
+    return sorted(known_loggers, key=lambda x: '' if not x else x)
+
+#----------------------------- from standart logging module width some modifications ----------------------------
+class ethLogger(logging.Logger):
+    def __init__(self, name, level=DEFAULT_LOGLEVEL):
+        super(ethLogger, self).__init__(name, level=level)
+
+    def is_active(self, level_name='trace'):
+        return self.isEnabledFor(logging._checkLevel(level_name.upper()))
+        #getLogger()
+        #self.setLevel(getattr(logging, level_name.upper()))
 
 
+class RootLogger(ethLogger):
+    """
+    A root logger is not that different to any other logger, except that
+    it must have a logging level and there is only one instance of it in
+    the hierarchy.
+    """
+    def __init__(self, level, log_json=False):
+        """
+        Initialize the logger with the name "root".
+        """
+        ethLogger.__init__(self, "root", level)
+        self.log_json = False
+
+class ethManager(logging.Manager):
+    def __init__(self, rootnode):
+        self.loggerClass = ethLogger
+        super(ethManager, self).__init__(rootnode)
+
+    def getLogger(self, name):
+        logging.setLoggerClass(ethLogger)
+        return super(ethManager, self).getLogger(name)
+
+rootLogger = RootLogger(DEFAULT_LOGLEVEL)
+ethLogger.root = rootLogger
+ethLogger.manager = ethManager(ethLogger.root)
+
+def help_make_kws(kws, name, msg, log_json=False):
+    new_kws = {}
+    message = ""
+
+    if log_json == True:
+        message = {}
+        for i in kws:
+            if i not in formatter_list:  # levelno, asctime......
+                print i, "=", kws[i]
+                #s = " {i}={kwsi}".format(i=i, kwsi=str(kws[i]))
+                #message += s
+                #message[i] = repr(kws[i]) if type(kws[i]).__name__=='instance'   else kws[i]
+                #if type(kws[i]) == (types.IntType or types.LongType or types.FloatType or types.ComplexType or types.DictionaryType):
+                #    message[i] = kws[i] # for json dumps
+                #else:
+                #    message[i] = repr(kws[i]) # for json dumps
+                kws_i_d = getattr(kws[i], "__dict__", False)
+                if kws_i_d != False:
+                    #message[i] = "{} {}".format(repr(kws[i]), kws[i].__dict__)
+                    message[i] = "{}".format(kws[i].__dict__)
+                else:
+                    message[i] = kws[i]
+            else:
+                new_kws[i] = kws[i]
+    else:
+        message = ""
+        for i in kws:
+            if i not in formatter_list:
+                print i, "=", kws[i]
+                s = " {i}={kwsi}".format(i=i, kwsi=str(kws[i]))
+                message += s
```

Comment:
Strings should not be built by repeatedly appending to them with the `+` operator. Instead append to a list and `''.join(messages)` at the end. See: https://docs.python.org/3/faq/programming.html#what-is-the-most-efficient-way-to-concatenate-many-strings-together